Ical.Net/DataTypes/Duration.cs

```diff
@@ -0,0 +1,222 @@
+    public static Duration Parse(string value) => (Duration)new DurationSerializer().Deserialize(new StringReader(value));
+
+    /// <summary>
+    /// Creates an instance that represents the given time span as exact value, that is, time-only.
+    /// </summary>
+    /// <remarks>
+    /// According to RFC5545 the weeks and day fields of a duration are considered nominal durations while the time fields are considered exact values.
+    /// </remarks>
+    internal static Duration FromTimeSpanExact(TimeSpan t)
+        // As a TimeSpan always refers to exact time, we specify days as part of the hours field,
+        // because time is added as exact values rather than nominal according to RFC 5545.
+        => new Duration(hours: NullIfZero(t.Days * 24 + t.Hours), minutes: NullIfZero(t.Minutes), seconds: NullIfZero(t.Seconds));
+
+    /// <summary>
+    /// Creates an instance that represents the given time span, treating the days as nominal duration and the time part as exact.
+    /// </summary>
+    /// <remarks>
+    /// According to RFC5545 the weeks and day fields of a duration are considered nominal durations while the time fields are considered exact values.
+    /// </remarks>
+    internal static Duration FromTimeSpan(TimeSpan t)
+        => new Duration(days: NullIfZero(t.Days), hours: NullIfZero(t.Hours), minutes: NullIfZero(t.Minutes), seconds: NullIfZero(t.Seconds));
+
+    /// <summary>
+    /// Gets a value representing the time parts of the given instance.
+    /// </summary>
+    internal TimeSpan TimeAsTimeSpan
+    {
+        get
+        {
+            return new TimeSpan(
+                0,
+                Hours ?? 0,
+                Minutes ?? 0,
+                Seconds ?? 0);
+        }
+    }
+
+    /// <summary>
+    /// Gets a value representing the date parts (days and weeks) of the given instance.
+    /// </summary>
+    internal TimeSpan DateAsTimeSpan
+    {
+        get
+        {
+            return new TimeSpan(
+                (Weeks ?? 0) * 7 + (Days ?? 0),
+                0,
+                0,
+                0);
+        }
+    }
+
+    /// <summary>
+    /// Convert the instance to a <see cref="TimeSpan"/>.
+    /// </summary>
+    internal TimeSpan ToTimeSpan()
+        => new TimeSpan(
+            (Weeks ?? 0) * 7 + (Days ?? 0),
+            Hours ?? 0,
+            Minutes ?? 0,
+            Seconds ?? 0);
+
+    /// <summary>
+    /// Gets a value indicating whether the duration is zero, that is, all fields are null or 0.
+    /// </summary>
+    internal bool IsZero
+        => ((Weeks ?? 0) == 0)
+            && ((Days ?? 0) == 0)
+            && ((Hours ?? 0) == 0)
+            && ((Minutes ?? 0) == 0)
+            && ((Seconds ?? 0) == 0);
+
+    /// <summary>
+    /// Gets a value indicating whether the duration is empty, that is, all fields are null.
+    /// </summary>
+    internal bool IsEmpty
```

Comment:
Would `Duration? x = null; if (x == null)` be more clear/simple than than `Duration x = new Duration(); if (x.IsEmpty)`?

Reply:
> Would Duration? x = null; if (x == null) be more clear/simple than than Duration x = new Duration(); if (x.IsEmpty)?

Basically yes, but we don't use `new Duration()` like that anyways. The only place we use it is to initialize `public static Duration Zero { get; } = new Duration();`. Not sure whether it would be desirable to disallow an empty Duration. Would rather say we should allow it. TBD.

Reply:
Oh yes, right, then `static Duration Zero { get; } = new Duration();` returns a duration that `IsEmpty`, not `IsZero`?

Reply:
Actually it is both, empty and zero.

I.e. `new Duration()` is both, empty and zero,
`new Duration(0)` is zero but not empty.

Both are `internal`, as I'm aware that it could be confusing.